Add basic timezone support for form datetime input.

dereuromark opened the pull request:
Basic stab at https://github.com/cakephp/cakephp/issues/10877

Currently all forms are broken if you follow recommended UTC for DB and localized forms (e.g. UTC+1 Berlin time).
The time in forms is always behind upon transformation to output, and when you save it, it becomes even more behind as it converts it wrong a 2nd time.
Using a defaultOutputTimezone config (and optionally even a timezone dropdown maybe) you can have localized forms and keep the expected timezoned datetime.

If there is only one timezone (output), one can just use config here:
```php
	'App' => [
		'defaultLocale' => env('APP_DEFAULT_LOCALE', 'de_DE'),
		'defaultOutputTimezone' => env('APP_DEFAULT_OUTPUT_TIMEZONE', 'Europe/Berlin'),
```

I kindly ask for help to finalize this for 3.6.

dereuromark committed:
Add basic timezone support for form datetime input.

dereuromark committed:
Fix CS.

dereuromark committed:
Fix typehint for phpstan.

dereuromark committed:
Fix doc blocks.

saeideng commented:
looks this is only for showing , not converting request data from user

dereuromark commented:
Both, it converts correctly for display in the input form (dropdowns by default), and it converts back before validating/saving.
Not yet sure about any edge cases yet though.
But seems BC so far, if you do not use any timezoning here.

saeideng commented:
do you think we can add calender in future ?(non-Gregorian)

saeideng commented:
how happen if someone remove hidden timezone field ?

markstory commented on src/View/Widget/DateTimeWidget.php:
```diff
@@ -93,6 +94,7 @@ public function __construct(StringTemplate $templates, SelectBoxWidget $selectBo
      * - `second` - Set to true to enable the seconds input. Defaults to false.
      * - `meridian` - Set to true to enable the meridian input. Defaults to false.
      *   The meridian will be enabled automatically if you choose a 12 hour format.
+     * - `timezone` - A hidden field for timezone handling when working with UTC internally.
```
Not just UTC really. Any timezone could be stored in the hidden input.

dereuromark replied:
The point is that you store anything but UTC usually :) The UTC is the server timezone for DB storing it normalized, and the timezone is either passed in the array from user input (session data?) or coming from config by default (e.g. german based website using global gmt+1 output).
So the idea is UTC => GMT+1 for display => UTC for form handling.
Thus "working with UTC internally" - which seems to be the best practice when dealing with localized datetime across multiple timezones.

markstory replied:
I agree UTC is what we should recommend, but people will do all sorts of interesting things.

markstory commented on src/View/Widget/DateTimeWidget.php:
```diff
@@ -528,6 +540,26 @@ protected function _meridianSelect($options, $context)
+            'templateVars' => [],
+        ];
+
+        return $this->_select->render($options, $context);
```
Why are we making a select and not a hidden input? Seems like a hidden input is what we want here.

dereuromark replied:
I agree. How would that look like here? Using a widget in a widget?

markstory replied:
That's what `$this->_select` is as well. The WidgetRegistry stitches various widget dependencies together.

markstory commented on src/Database/Type/DateTimeType.php:
```diff
@@ -206,9 +207,17 @@ public function marshal($value)
+        $internalTimezone = date_default_timezone_get();
+        if ($tz && $tz !== $internalTimezone) {
+            $time = $time->timezone($internalTimezone);
+        }
```
Does the time type need this kind of change as well?

dereuromark replied:
I would think so, so far I only checked datetime. If anyone has the chance to help with other types, that would be more than welcome.

ADmad commented on src/View/Widget/DateTimeWidget.php:
```diff
@@ -211,25 +214,30 @@ protected function _normalizeData($data)
+                $date = new DateTime('now', $timezone);
             } elseif (is_int($value) || is_numeric($value)) {
-                $date = new DateTime('@' . $value);
+                $date = new DateTime('@' . $value, $timezone);
```
Passing `$timezone` here is redundant. From the PHP manual:
> The $timezone parameter and the current timezone are ignored when the $time parameter either is a UNIX timestamp (e.g. @946684800) or specifies a timezone (e.g. 2010-01-28T15:00:00+02:00).

ADmad commented on src/Database/Type/DateTimeType.php:
```diff
@@ -206,9 +207,17 @@ public function marshal($value)
+
+        $internalTimezone = date_default_timezone_get();
+        if ($tz && $tz !== $internalTimezone) {
+            $time = $time->timezone($internalTimezone);
```
Doesn't this change mean the timezone should be again set to default timezone before converting the object to string in `toDatabase()` method?

dereuromark replied:
My tests seemed to yield the right conversion, so maybe we do not? But I am happy to be proven wrong.

josegonzalez committed:
docs: switch the UTC reference

josegonzalez committed:
chore: drop extra parameter

It is unused for unix timestmaps.

ADmad committed:
Generate hidden input for timezone.

Skip hidden input generation if default output timezone is not set
or explicitly disabled through widget data.

dereuromark commented:
Anyone able to help out, getting this ready for 3.7 maybe?
We could also look into https://github.com/JeanValJeann/cakephp-timezone and what they did.
